in_toto/verifylib.go

```diff
@@ -0,0 +1,517 @@
+                  " don't support rules of type '%s'", rule,
+                  strings.ToUpper(ruleData["type"]))
+        }
+      }
+    }
+  }
+  return nil
+}
+
+
+func ReduceStepsMetadata(layout Layout,
+    stepsMetadata map[string]map[string]Metablock) (map[string]Metablock,
+    error) {
+  stepsMetadataReduced := make(map[string]Metablock)
+
+  for _, step := range layout.Steps {
+    linksPerStep, ok := stepsMetadata[step.Name]
+    // We should never get here, layout verification must fail earlier
+    if !ok || len(linksPerStep) < 1 {
+      panic("Could not reduce metadata for step '" + step.Name +
+          "', no link metadata found.")
+    }
+
+    // Get the first link (could be any link) for the current step, which will
+    // serve as reference link for below comparisons
+    var referenceKeyID string
+    var referenceLinkMb Metablock
+    for keyID, linkMb := range linksPerStep {
+      referenceLinkMb = linkMb
+      referenceKeyID = keyID
+      break
+    }
+
+    // Only one link, nothing to reduce, take the reference link
+    if len(linksPerStep) == 1 {
+      stepsMetadataReduced[step.Name] = referenceLinkMb
+
+    // Multiple links, reduce but first check
+    } else {
+      // Artifact maps must be equal for each type among all links
+      // TODO: What should we do if there are more links, than the
+      // threshold requires, but not all of them are equal? Right now we would
+      // also error.
+      for keyID, linkMb := range linksPerStep {
+        if !reflect.DeepEqual(linkMb.Signed.(Link).Materials,
+            referenceLinkMb.Signed.(Link).Materials) ||
+            !reflect.DeepEqual(linkMb.Signed.(Link).Products,
+            referenceLinkMb.Signed.(Link).Products) {
+          return nil, fmt.Errorf("Link '%s' and '%s' have different" +
+              " artifacts.",
+              fmt.Sprintf(LinkNameFormat, step.Name, referenceKeyID),
+              fmt.Sprintf(LinkNameFormat, step.Name, keyID))
+        }
+      }
+      // We haven't errored out, so we can reduce
+      stepsMetadataReduced[step.Name] = referenceLinkMb
+    }
+  }
+  return stepsMetadataReduced, nil
+}
+
+
+func VerifyStepCommandAlignment(layout Layout,
+    stepsMetadata map[string]map[string]Metablock) {
+  for _, step := range layout.Steps {
+    linksPerStep, ok := stepsMetadata[step.Name]
+    // We should never get here, layout verification must fail earlier
+    if !ok || len(linksPerStep) < 1 {
+      panic("Could not verify command alignment for step '" + step.Name +
+          "', no link metadata found.")
+    }
+
+    for signerKeyID, linkMb := range linksPerStep {
+      expectedCommandS := strings.Join(step.ExpectedCommand, " ")
+      executedCommandS := strings.Join(linkMb.Signed.(Link).Command, " ")
+
+      if expectedCommandS != executedCommandS {
+        linkName := fmt.Sprintf(LinkNameFormat, step.Name, signerKeyID)
+        fmt.Printf("WARNING: Expected command for step '%s' (%s) and command" +
```

Comment:
I wonder if we want to print to stderr instead.

Reply:
Yes! I wanted to rework user output and logging anyway. Haven't figured out yet how to best do this in go. :) Is it okay, if I create a ticket for this?

Reply:
See https://github.com/in-toto/in-toto-golang/issues/4